src/kvstore/collectives/include/coll_wrapper.h

```diff
@@ -0,0 +1,114 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * Copyright (c) 2018 by Contributors
+ */
+
+#ifndef MXNET_KVSTORE_COLLECTIVES_INCLUDE_COLL_WRAPPER_H_
+#define MXNET_KVSTORE_COLLECTIVES_INCLUDE_COLL_WRAPPER_H_
+
+#if MXNET_USE_ALLREDUCE_DIST_KVSTORE
+
+#include <mpi.h>
+
+#include "mxnet/ndarray.h"
+#include "mxnet/base.h"
+#include "mpi_message.pb.h"
+
+template<typename DType>
+MPI_Datatype MPI_Data_Type_Cast(void);
+
+template<>
+MPI_Datatype MPI_Data_Type_Cast<int>(void) {
+  return MPI_INT;
+}
+
+template<>
+MPI_Datatype MPI_Data_Type_Cast<float>(void) {
+  return MPI_FLOAT;
+}
+
+template<>
+MPI_Datatype MPI_Data_Type_Cast<double>(void) {
+  return MPI_DOUBLE;
+}
+
+template <class xpu, class DType>
+struct COLL_Wrapper {
+  static int Broadcast(mxnet::NDArray *input_array,
+                       int root_rank) {
+    return 0; }
+
+  static int AllReduce(mxnet::NDArray *input_array,
+                       mxnet::NDArray *output_array) {
+    return 0; }
+};
+
+// CPU Implementation
+template <class DType>
+struct COLL_Wrapper<mxnet::cpu, DType> {
+  static int Broadcast(mxnet::NDArray *input_array,
+                       int root_rank) {
+    DType *buf = reinterpret_cast<DType *>(input_array->data().dptr<DType>());
+    unsigned int count = input_array->data().Size();
+    int ret = MPI_Bcast(buf, count, MPI_Data_Type_Cast<DType>(), root_rank, MPI_COMM_WORLD);
+    return ret;
+  }
+
+  static int AllReduce(mxnet::NDArray *input_array,
+                       mxnet::NDArray *output_array) {
+    DType *send_buf = reinterpret_cast<DType *>(input_array->data().dptr<DType>());
+    DType *recv_buf = reinterpret_cast<DType *>(output_array->data().dptr<DType>());
+    unsigned int count = input_array->data().Size();
+    int ret;
+    assert(input_array->data().Size() == output_array->data().Size());
+
+    if (send_buf != recv_buf) {
+      ret = MPI_Allreduce(reinterpret_cast<const void *>(send_buf),
+                          reinterpret_cast<void *>(recv_buf),
+                          count, MPI_Data_Type_Cast<DType>(), MPI_SUM, MPI_COMM_WORLD);
+    } else {
+      ret = MPI_Allreduce(MPI_IN_PLACE, reinterpret_cast<void *>(recv_buf),
+                         count, MPI_Data_Type_Cast<DType>(), MPI_SUM, MPI_COMM_WORLD);
+    }
+    return ret;
+  }
+};
+
+// GPU Implementation
+template <class DType>
+struct COLL_Wrapper<mxnet::gpu, DType> {
+  static int Broadcast(mxnet::NDArray *input_array,
+                       int root_rank) {
+    // TODO(zhouhaiy): implement gpu broadcast
+    LOG(FATAL) << "Collective For GPU version has not been implemented.";
```

Comment:
From what I understand this is pretty easy to fill and enhance the functionality of this PR by adding GPU support. By using the CommDevice class in comm.h. You just need to call Reduce function from that class. Isn't it?

Reply:
It's not that easy, though overall it works as your think.
Reduce function shouldn't be called here. 
It should be called after entering MXAllReduce. 
1) GroupKVPairs CommDevice::Reduce (MXAllReduce)
2) same as CPU logic
3) CommDevice::Broadcast (PerformCollectiveOp)
Compared with CPU, GPU path has extra 1) and 3).